compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp

```diff
@@ -521,6 +521,76 @@ static LogicalResult setWarpReductionConfig(func::FuncOp entryPoint,
   return success();
 }
 
+/// Returns true if the operation is a GenericOp implementing a 2D transpose.
+static bool isTransposeOp(linalg::LinalgOp linalgOp) {
```

Comment:
Just FYI, this is going to be a really really limited case. It is extremely rare you will get a transpose that is just

```
%1 = linalg.generic {
   indexing_maps = [affine_map<(d0, d1) -> (d0, d1)>, affine_map<(d0, d1) -> (d1, d0)>],
   iterator_types = ["parallel", "parallel"]}
   ins(%0 : tensor<?x?xf32>) outs(%init : tensor<?x?xf32>)
  ^bb0(%arg0 : f32, %arg1 : f32):
    linalg.yield %arg0 : f32
} -> tensor<?x?xf32>
```

So I think this is really limited. I dont know why this is being added (if it is just a strawman thing, or part of something that is to be generalized), but it would be extremely rare for anything in a real model to hit this.

Reply:
yes the goal is to generalize it to more complex transpose and to enable it with fusion with elementwise, reductions, etc... (this is mentioned in the issue attached)
Considering this is not a trivial change we want to go by step.

Reply:
Understood. But adding a new pipeline seems like a big hammer to start with....

Reply:
It looked like the simplest solution to me since it allows us to land code with a small scope to start with. How would you do it?

Reply:
I would not focus on a transpose e2e test but rather just start with a test case that is already distributed so that you are doing transpose for a tile. Lower that without having to mess with KernelConfig, etc. Maybe even have hard coded configs . Cause what you really are looking for AFAICS is the actual transformation that loads things into shared memory and then lowers it to you code. The e2e test is less interesting IMO

Reply:
To be able to do the tiling efficiently we need a special tile size. If we use the current default tile size we cannot do the code generation we want. Most of the stuff we want to do is pick a good tile size and make it work well with shared memory copy and then avoid bank conflicts. This is mostly handling the e2e flow.

Maybe I'm missing your point. Let's have a quick chat if you have concerns.

Reply:
Just looking at what is added here these feels like a special pipeline for transpose op by itself. My only point is that is very short term and just questioning adding that (technical) overhead for something that is short term.
Had a quick offline chat. I am not blocking the PR (havent yet). Just doing the "see something say something" but this seems to be an transient phase, so not blocking this PR (havent blocked it now either)

Reply:
@MaheshRavishankar do you have examples of other kinds of transpose this should apply to.  The cases I'm looking at expanding this to are:

- Transpose more than 2 dimensions and support arbitrary permutations
- Transpose fused with other element wise

Anything else missing?

Reply:
That covers it, I think. Essentially its any generic op where the indexing maps are (projected) permutations. So the strategy here should only rely on those (you should not have to look into the region of the generic op, for example).

Reply:
Looking at the code now, you dont do that. So thats great. Eventually it'd be great if there is no `isTransposeOp` or `setTransposeConfig`, but rather the same heuristics for all `linalg.generic` ops, based on analysis of the indexing maps.